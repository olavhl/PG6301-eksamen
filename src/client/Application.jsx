import React from "react";
import { BrowserRouter, Link } from "react-router-dom";
import { Route, Switch } from "react-router";

export function Application() {
  return (
    <BrowserRouter>
      <header>
        <Link to={"/"}>
          <button>Home</button>
        </Link>
      </header>
      <Switch>
        <Route exact path={"/login"}>
          <h1>Login</h1>
        </Route>
        <Route path={"/login/callback"}>
          <h1>Login Callback</h1>
        </Route>
        <Route exact path={"/"}>
          <h1>Welcome to the chatting page</h1>
          <ul>
            <li>
              <Link to={"/login"}>Login</Link>
            </li>
          </ul>
        </Route>
        <Route>
          <h1>Page not found...</h1>
        </Route>
      </Switch>
    </BrowserRouter>
  );
}
