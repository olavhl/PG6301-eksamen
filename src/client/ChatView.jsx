import React, { useState } from "react";

export function ChatView({ onSendMessage, chat }) {
  const [message, setMessage] = useState("");
  // Feedback
  const [feedBack, setFeedBack] = useState("");

  function handleSubmitChat(e) {
    e.preventDefault();

    if (message === "") {
      setFeedBack("Write the message you want to send!");
      return;
    }

    onSendMessage(message);
    setMessage("");
  }

  return (
    <div className={"chat"}>
      <header>
        <h1>Chat</h1>
      </header>
      <main>
        {chat?.map((message, index) => (
          <div key={`${message}__${index}`}>{message}</div>
        ))}
      </main>
      {feedBack && <div className={"feedback"}>{feedBack}</div>}
      <footer>
        <form onSubmit={handleSubmitChat}>
          <input
            type="text"
            autoFocus={true}
            value={message}
            onChange={(e) => setMessage(e.target.value)}
          />
          <button>Send</button>
        </form>
      </footer>
    </div>
  );
}
