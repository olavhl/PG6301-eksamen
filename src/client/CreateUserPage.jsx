import React, { useState } from "react";
import { InputField } from "./components/InputField";
import { useHistory } from "react-router";

export function CreateUserPage({ userApi }) {
  const [firstName, setFirstName] = useState("");
  const [lastName, setLastName] = useState("");
  const [email, setEmail] = useState("");
  // Feedback
  const [feedBack, setFeedBack] = useState("");

  const history = useHistory();

  async function handleSubmit(e) {
    e.preventDefault();

    if (firstName === "" || lastName === "" || email === "") {
      setFeedBack(
        "You need to implement First and Last name, as well as email"
      );
      return;
    }

    await userApi.createUser({ firstName, lastName, email });
    history.push("/user/list");
  }

  return (
    <div className={"cont"}>
      <h1>Create user</h1>
      <form onSubmit={handleSubmit}>
        <InputField
          value={firstName}
          label={"First name"}
          onChangeValue={setFirstName}
        />
        <InputField
          value={lastName}
          label={"Last name"}
          onChangeValue={setLastName}
        />
        <InputField value={email} label={"Email"} onChangeValue={setEmail} />
        <button>Submit</button>
      </form>
      <br />
      {feedBack && <div className={"feedback"}>{feedBack}</div>}
    </div>
  );
}
