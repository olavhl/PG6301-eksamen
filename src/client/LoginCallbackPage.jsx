import React, { useEffect } from "react";
import { useHistory } from "react-router";

export function LoginCallbackPage({ onAccessToken }) {
  const hash = Object.fromEntries(
    new URLSearchParams(window.location.hash.substring(1))
  );

  const history = useHistory();

  useEffect(() => {
    const { access_token } = hash;
    onAccessToken(access_token);
    history.push("/");
  }, []);

  return <h1>Login Callback</h1>;
}
