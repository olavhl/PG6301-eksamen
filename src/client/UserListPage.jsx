import React from "react";
import { useLoading } from "./lib/useLoading";
import { ErrorView } from "./components/ErrorView";
import { LoadingView } from "./components/LoadingView";
import { UserCreateUserButton } from "./components/UserCreateUserButton";
import { Link } from "react-router-dom";

export function UserListPage({ userApi }) {
  const { data: users, error, loading, reload } = useLoading(
    async () => await userApi.listUsers()
  );

  if (error) {
    return <ErrorView error={error} reload={reload} />;
  }

  if (loading) {
    return <LoadingView />;
  }

  // Checking if there are any users in the UserArray in server
  if (users.length === 0) {
    return <UserCreateUserButton />;
  }

  return (
    <div>
      <h1>All Users</h1>
      <Link to={"/user/create"}>
        <button>Create user</button>
      </Link>
      {users?.map(({ firstName, lastName, email, id }) => (
        <div key={`${firstName}__${id}`}>
          <p>Firstname: {firstName}</p>
          <p>Lastname: {lastName}</p>
          <p>Email: {email}</p>
          <Link to={`/user/${id}/message`}>
            <button>Send Message</button>
          </Link>
          <br />
        </div>
      ))}
    </div>
  );
}
