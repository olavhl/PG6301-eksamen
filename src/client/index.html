<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <title>Messaging System</title>
    <link rel="stylesheet" href="style.css" type="text/css">
  </head>
  <body>
    <div id="app"></div>
  </body>
  <script src="index.jsx"></script>
</html>
