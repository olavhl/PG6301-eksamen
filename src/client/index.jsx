import React from 'react';
import ReactDOM from 'react-dom';

function Application() {
    return <h1>Hello</h1>;
}

ReactDOM.render(<Application/>, document.getElementById("app"));