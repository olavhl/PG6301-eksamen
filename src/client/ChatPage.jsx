import React, { useEffect, useState } from "react";
import { ChatView } from "./ChatView";

export function ChatPage() {
  const [chat, setChat] = useState([]);
  const [ws, setWs] = useState();

  useEffect(() => {
    const ws = new WebSocket("ws://localhost:3000/chat");
    ws.onopen = (event) => {
      console.log("opened", event);
    };

    ws.onmessage = (event) => {
      console.log("message", event);
      setChat((chat) => [...chat, event.data]);
    };

    ws.onclose = (event) => {
      console.log("close", event);
    };

    setWs(ws);
  }, []);

  return <ChatView chat={chat} onSendMessage={(message) => ws.send(message)} />;
}
