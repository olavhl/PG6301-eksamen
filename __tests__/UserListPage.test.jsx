import React from "react";
import ReactDOM from "react-dom";
import { act } from "react-dom/test-utils";
import { MemoryRouter } from "react-router";
import { UserListPage } from "../src/client/UserListPage";

async function renderToDOM(component) {
  const container = document.createElement("div");
  document.body.appendChild(container);
  await act(async () => {
    ReactDOM.render(<MemoryRouter>{component}</MemoryRouter>, container);
  });
  return container;
}

describe("user list page", () => {
  it("shows users", async () => {
    const listUsers = async () => [
      {
        id: 1,
        firstName: "Ola",
        lastName: "Nordmann",
        email: "ola@nordmann.no",
      },
    ];

    const container = await renderToDOM(
      <UserListPage userApi={{ listUsers }} />
    );

    expect(container.innerHTML).toMatchSnapshot();
    expect(container.querySelector("p").textContent).toEqual("Firstname: Ola");
  });

  it("shows loading screen", async () => {
    const listUsers = () => new Promise(() => {});
    const container = await renderToDOM(
      <UserListPage userApi={{ listUsers }} />
    );

    expect(container.innerHTML).toMatchSnapshot();
    expect(container.querySelector("div").textContent).toEqual("Loading...");
  });

  it("shows error screen", async () => {
    const listUsers = async () => {
      throw Error("This test should show an Error..");
    };
    const container = await renderToDOM(
      <UserListPage userApi={{ listUsers }} />
    );

    expect(container.innerHTML).toMatchSnapshot();
    expect(container.querySelector("div").textContent).toEqual(
      "Something went wrongError: This test should show an Error.."
    );
  });

  it("shows no users-button", async () => {
    const listUsers = () => [];
    const container = await renderToDOM(
      <UserListPage userApi={{ listUsers }} />
    );

    expect(container.innerHTML).toMatchSnapshot();
    expect(container.querySelector("h3").textContent).toEqual(
      "You dont have any users/friends :("
    );
  });
});
