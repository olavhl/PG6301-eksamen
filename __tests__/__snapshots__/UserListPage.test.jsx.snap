// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user list page shows error screen 1`] = `"<div class=\\"cont\\"><div>Something went wrong<br>Error: This test should show an Error..</div><button>Try again</button></div>"`;

exports[`user list page shows loading screen 1`] = `"<div class=\\"cont\\">Loading...</div>"`;

exports[`user list page shows no users-button 1`] = `"<div class=\\"cont\\"><h3>You dont have any users/friends :(</h3><a href=\\"/user/create\\"><button>Create user</button></a></div>"`;

exports[`user list page shows users 1`] = `"<div class=\\"cont\\"><h1>All Users</h1><a href=\\"/user/create\\"><button>Create user</button></a><div><p>Firstname: Ola</p><p>Lastname: Nordmann</p><p>Email: ola@nordmann.no</p><a href=\\"/user/1/message\\"><button>Send Message</button></a><br></div></div>"`;
